File: output/sonarqube_analysis/copilot_solutions/python/results/q28-number-of-islands.py
# https://leetcode.com/problems/number-of-islands/
# Problem: Number of Islands 

#Problem Description:
#Given an m x n 2D binary grid grid which represents a map of '1's (land) and '0's (water), return the number of islands.
#An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.

# Constraints:
# m == grid.length
# n == grid[i].length
# 1 <= m, n <= 300
# grid[i][j] is '0' or '1'.

#Code Structure:
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if not grid:
            return 0
        
        rows, cols = len(grid), len(grid[0])
        visited = [[False] * cols for _ in range(rows)]
        
        def dfs(r, c):
            if r < 0 or r >= rows or c < 0 or c >= cols or grid[r][c] == '0' or visited[r][c]:
                return
            visited[r][c] = True
            # Explore all four directions
            dfs(r + 1, c)
            dfs(r - 1, c)
            dfs(r, c + 1)
            dfs(r, c - 1)
        
        island_count = 0
        
        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == '1' and not visited[i][j]:
                    dfs(i, j)
                    island_count += 1
        
        return island_count                                                                                                   
    

File: output/sonarqube_analysis/copilot_solutions/php/results/q28-number-of-islands.php
<?php
/** https://leetcode.com/problems/number-of-islands */
// Problem: Number of Islands 

//Problem Description:
//Given an m x n 2D binary grid grid which represents a map of '1's (land) and '0's (water), return the number of islands.
//An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.

// Constraints:
// m == grid.length
// n == grid[i].length
// 1 <= m, n <= 300
// grid[i][j] is '0' or '1'.

//Code Structure:
class Solution {

    /**
     * @param String[][] $grid
     * @return Integer
     */
    function numIslands($grid) {
        if (empty($grid)) {
            return 0;
        }

        $rows = count($grid);
        $cols = count($grid[0]);
        $islandCount = 0;

        for ($i = 0; $i < $rows; $i++) {
            for ($j = 0; $j < $cols; $j++) {
                if ($grid[$i][$j] === '1') {
                    $this->dfs($grid, $i, $j, $rows, $cols);
                    $islandCount++;
                }
            }
        }

        return $islandCount;  
    }

    private function dfs(&$grid, $i, $j, $rows, $cols) {
        if ($i < 0 || $i >= $rows || $j < 0 || $j >= $cols || $grid[$i][$j] === '0') {
            return;
        }

        // Mark the land as visited
        $grid[$i][$j] = '0';

        // Explore all four directions
        $this->dfs($grid, $i + 1, $j, $rows, $cols); // Down
        $this->dfs($grid, $i - 1, $j, $rows, $cols); // Up
        $this->dfs($grid, $i, $j + 1, $rows, $cols); // Right
        $this->dfs($grid, $i, $j - 1, $rows, $cols); // Left
    }
}
    